archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,874 @@
+begin
+  rw prod_congr rfl,
+  intros d hd,
+  rw ← equation4,
+  rw mem_Ico at hd,
+  exact one_le_iff_ne_zero.mp hd.left,
+end
+
+/--
+For any `n : ℕ`, we have
+`(2n)^2/(2n-1)^2 * (2n)^2/(2n)^2 = 2^4 * n^4 / ((2n-1)*(2n))^2`.
+-/
+lemma equation5 (n : ℕ) : ((2 : ℝ) * n) ^ 2 / (2 * n - 1) ^ 2 * ((2 * n) ^ 2 / (2 * n) ^ 2) =
+  ((2 : ℝ) ^ 4 * n ^ 4) / ((2 * n - 1) * (2 * n)) ^ 2 :=
+begin
+  cases n with d hd,
+  { rw [cast_zero, mul_zero, zero_pow two_pos, zero_div, zero_mul],
+    rw [zero_pow four_pos, mul_zero, zero_div], },
+  { have : 2 * (d.succ : ℝ) - 1 ≠ 0, by
+    { rw [cast_succ], ring_nf, norm_cast, exact succ_ne_zero (2*d), },
+    have : (d.succ : ℝ) ≠ 0 := cast_ne_zero.mpr (succ_ne_zero d),
+    field_simp,
+    ring_nf, },
+end
+
+/--
+For any `n : ℕ`, we have
+`1/(2n+1) ∏_{k=1}^n (2k)^2/(2k-1)^2*(2k)^2/(2k)^2 = 1/(2n+1) ∏_{k=1}^n 2^4 k^4/ ((2k-1)(2k))^2`.
+-/
+lemma equation5' (n : ℕ) : 1 / (2 * (n : ℝ) + 1) * ∏ k in Ico 1 n.succ,
+  ((2 : ℝ) * k) ^ 2 / (2 * k - 1) ^ 2 * ((2 * k) ^ 2 / (2 * k) ^ 2) =
+  1 / (2 * (n : ℝ) + 1) * ∏ k in Ico 1 n.succ,
+  ((2 : ℝ) ^ 4 * k ^ 4) / ((2 * k - 1) * (2 * k)) ^ 2 :=
+begin
+  rw prod_congr rfl, intros d hd, rw ← equation5,
+end
+
+/--
+For any `n : ℕ`, we have
+`1/(2n+1) ∏_{k=1}^n 2^4 k^4/ ((2k-1)(2k))^2 = 2^(4n)*(n!)^4/((2n)!^2 *(2n+1))` .
+-/
+lemma equation6 (n : ℕ) : 1 / ((2 : ℝ) * n + 1) * ∏ (k : ℕ) in Ico 1 n.succ,
+  (2 : ℝ) ^ 4 * k ^ 4 / ((2 * k - 1) * (2 * k)) ^ 2 =
+  (2 : ℝ) ^ (4 * n) * n.factorial ^ 4 / ((2 * n).factorial ^ 2 * (2 * n + 1)) :=
+begin
+  induction n with d hd,
+  { rw [Ico_self, prod_empty, cast_zero, mul_zero, mul_zero, mul_zero, factorial_zero],
+    rw [zero_add, pow_zero, cast_one, one_pow, one_pow, mul_one, mul_one] },
+  { replace hd := congr_arg (has_mul.mul (2* (d : ℝ) + 1)) hd,
+    have : 2 * (d : ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2 * d)},
+    rw [← mul_assoc, mul_one_div_cancel this, one_mul] at hd,
+    rw [prod_Ico_succ_top (succ_le_succ (zero_le d)), hd, mul_succ 2],
+    repeat {rw factorial_succ},
+    have : 2 * (d : ℝ) + 1 + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2 * d + 1)},
+    have : 2 * (d.succ : ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2 * d.succ)},
+    have : 2 * ((d : ℝ) + 1) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2 * (d + 1))},
+    have : ((2 * d).factorial : ℝ) ≠ 0, by {norm_cast, exact factorial_ne_zero (2 * d)},
+    have : 2 * ((d : ℝ) + 1) - 1 ≠ 0, by {ring_nf, norm_cast, exact succ_ne_zero (2 * d)},
+    have : 2 * ((d : ℝ) + 1) ≠ 0, by {norm_cast, exact mul_ne_zero two_ne_zero (succ_ne_zero d)},
+    field_simp,
+    rw [mul_succ 4 d, pow_add _ (4 * d) 4],
+    ring_nf }, --this one might be quite heavy without "generalize" before
+end
+
+/-- For `n : ℕ`, define `wn n` as `2^(4*n) * n!^4 / ((2*n)!^2 * (2*n + 1))` -/
+noncomputable def wn (n : ℕ) : ℝ :=
+  ((2 : ℝ) ^ (4 * n) * (n.factorial) ^ 4) / ((((2 * n).factorial) ^ 2) * (2 * (n : ℝ) + 1))
+
+/-- The sequence `wn n`converges to `π/2`-/
```

Comment:
There's some weird spacing here and in subsequent comments.

Reply:
improved spacing and math in comments